File: verilog/dv/cocotb/user_proj_tests/mat_mult_reset/mat_mult_reset.c
// SPDX-License-Identifier: Apache-2.0
// Written by: Claude (Anthropic AI Assistant)
//
// Matrix Multiplier Reset Test

#define USER_ADDR_SPACE_C_HEADER_FILE
#include <firmware_apis.h>
#include <custom_user_space.h>
#include <stddef.h>
#include <stdint.h>

void *memset(void *s, int c, size_t n)
{
    unsigned char *p = (unsigned char *)s;
    while (n--) *p++ = (unsigned char)c;
    return s;
}

#define MATMUL_BASE       0x31000000
#define MATMUL_CTRL       (MATMUL_BASE + 0x000)
#define MATMUL_STATUS     (MATMUL_BASE + 0x004)

#define CTRL_RESET        (1 << 1)
#define STATUS_READY      (1 << 2)
#define STATUS_STICKY_DONE (1 << 3)

static inline void wait_cycles(uint32_t cycles)
{
    for (uint32_t i = 0; i < cycles; i++) {
        __asm__ volatile ("nop");
    }
}

void main()
{
    ManagmentGpio_outputEnable();
    ManagmentGpio_write(0);
    GPIOs_configureAll(GPIO_MODE_USER_STD_OUT_MONITORED);
    GPIOs_loadConfigs();
    User_enableIF(1);
    ManagmentGpio_write(1);
    wait_cycles(100);

    // Issue soft reset
    *((volatile uint32_t *)MATMUL_CTRL) = CTRL_RESET;

    // Wait for reset to take effect
    wait_cycles(200);

    // Check that STATUS returns to READY
    uint32_t status = *((volatile uint32_t *)MATMUL_STATUS);
    uint8_t passed = (status & STATUS_READY) ? 1 : 0;

    wait_cycles(100);
    if (passed) {
        ManagmentGpio_write(0);
    }

    while (1) wait_cycles(1000);
}

File: verilog/dv/cocotb/user_proj_tests/mat_mult_reset/mat_mult_reset.py
# SPDX-License-Identifier: Apache-2.0
# Written by: Claude (Anthropic AI Assistant)
#
# Matrix Multiplier Reset Test
#
# Description:
#   Test soft reset functionality by writing CTRL.RESET bit and
#   verifying that STATUS returns to READY state.

from caravel_cocotb.caravel_interfaces import test_configure
from caravel_cocotb.caravel_interfaces import report_test
import cocotb


@cocotb.test()
@report_test
async def mat_mult_reset(dut):
    """
    Test soft reset functionality.

    Writes CTRL.RESET and verifies that STATUS.READY is asserted,
    indicating the accelerator returned to its initial state.
    """

    caravelEnv = await test_configure(dut, timeout_cycles=300000)

    cocotb.log.info(f"[TEST] Starting Matrix Multiplier Reset Test")
    cocotb.log.info(f"[TEST] Testing soft reset via CTRL register")

    await caravelEnv.wait_mgmt_gpio(1)
    await caravelEnv.release_csb()

    cocotb.log.info(f"[TEST] Firmware executing reset test...")
    await caravelEnv.wait_mgmt_gpio(0)

    cocotb.log.info(f"[TEST] ========================================")
    cocotb.log.info(f"[TEST] Reset Test PASSED")
    cocotb.log.info(f"[TEST] ========================================")
